traceevents/src/main/java/com/tomtom/kotlin/traceevents/TraceEventConsumerCollection.kt

```diff
@@ -32,33 +28,45 @@ import kotlin.reflect.jvm.javaMethod
  */
 class TraceEventConsumerCollection {
 
-    fun add(traceEventConsumer: TraceEventConsumer) {
-        traceEventsConsumers.add(traceEventConsumer)
+    fun add(traceEventConsumer: TraceEventConsumer, contextRegex: Regex?) {
+        traceEventsConsumersWithContext.add(TraceEventConsumerWithContext(traceEventConsumer, contextRegex))
     }
 
-    fun remove(traceEventConsumer: TraceEventConsumer) {
-        traceEventsConsumers.remove(traceEventConsumer)
+    fun remove(traceEventConsumer: TraceEventConsumer, contextRegex: Regex? = null) {
+        for (consumer in traceEventsConsumersWithContext.asIterable()
+            .filter {
+                it.traceEventConsumer == traceEventConsumer &&
+                        (contextRegex == null || it.contextRegex == contextRegex)
+            }) {
+            traceEventsConsumersWithContext.remove(consumer)
```

Comment:
Implementation can be simplified:

```
traceEventsConsumersWithContext.removeAll { <predicate> }
```

That would also prevent me from being worried about concurrent modification exceptions.

Reply:
Thanks... so much better. Apologies for being an old-style programmer.